src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,219 @@
+ * Therefore no need to go through the same set of validation steps/rules and just re-extract zip file and upload.
+ * In case some {@link StorageException} or {@link IOException} experienced during zip extraction
+ * the envelope state will not be changed and left for a retry on the next run.
+ * {@link Event#DOC_UPLOAD_FAILURE} can be treated as stuck envelope and be re-uploaded by same service here.
+ * There is an upload retry counter/limit which may be used as well
+ * Review ^ last sentence and, perhaps, incorporate such possibility here.
+ */
+@Service
+public class UploadEnvelopeDocumentsService {
+
+    private static final Logger log = getLogger(UploadEnvelopeDocumentsService.class);
+
+    private final BlobManager blobManager;
+    private final ZipFileProcessor zipFileProcessor;
+    private final DocumentProcessor documentProcessor;
+    private final EnvelopeProcessor envelopeProcessor;
+
+    public UploadEnvelopeDocumentsService(
+        BlobManager blobManager,
+        ZipFileProcessor zipFileProcessor,
+        DocumentProcessor documentProcessor,
+        EnvelopeProcessor envelopeProcessor
+    ) {
+        this.blobManager = blobManager;
+        this.zipFileProcessor = zipFileProcessor;
+        this.documentProcessor = documentProcessor;
+        this.envelopeProcessor = envelopeProcessor;
+    }
+
+    public void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
+
+        try {
+            CloudBlobContainer blobContainer = blobManager.getContainer(containerName);
+
+            envelopes.forEach(envelope -> processEnvelope(blobContainer, envelope));
+        } catch (URISyntaxException | StorageException exception) {
+            log.error("Unable to get client for {} container", containerName, exception);
+        }
+    }
+
+    private void processEnvelope(CloudBlobContainer blobContainer, Envelope envelope) {
+        String containerName = blobContainer.getName();
+        String zipFileName = envelope.getZipFileName();
+        UUID envelopeId = envelope.getId();
+        CloudBlockBlob blobClient = null;
+
+        try {
+            blobClient = getCloudBlockBlob(blobContainer, zipFileName, envelopeId);
+
+            Optional<String> lease = blobManager.acquireLease(blobClient, containerName, envelope.getZipFileName());
+
+            if (lease.isPresent()) {
+                BlobInputStream bis = getBlobInputStream(blobClient, containerName, zipFileName, envelopeId);
+                ZipFileProcessingResult result = processInputStream(bis, containerName, zipFileName, envelopeId);
+
+                uploadParsedZipFileName(envelope, result.getPdfs());
+
+                envelopeProcessor.handleEvent(envelope, DOC_UPLOADED);
+            }
+        } catch (UploadFailures failure) {
+            log.error(failure.getMessage(), failure.getCause());
```

Comment:
Individual specific exceptions are great - they make it easier to find out what happened and it's good we're throwing them. On this level I would still catch `Exception` here to be sure we covered everything (now and after future modifications, possibly faulty), and log with general error message. The cause will still result in all details being available. We will see that the job failed and we'll be able to find out why - both error messages will be available in logs.

Reply:
I can add generic one as well - no problem